### ISSUE TITLE
system message isn't being overridden when using the chat-completion API

### ISSUE THREAD

**Issue by jukofyork** [ollama/ollama#2470]
Sorry if this has been mentioned already (searching the Issues for "system" brings up 100s of pages):

```
{
  "model": "mixtral:32k-test",
  "messages": [
    {
      "role": "system",
      "content": "You are an AI assistant for the Eclipse IDE. Your objective is to assist users in writing and analyzing source code. Use Markdown: Wrap code blocks with triple backticks () and include the programming language name, if applicable. Use single backticks () to denote a word or phrase as code. Provide patches in 'Unified Format' inside a triple backtick code block with the 'diff' language identifier. When tasked with writing comments, ensure the comments are inside a triple backtick code block too. When tasked with writing code: 1. Understand the problem: constraints, specifications, objective, and edge cases. 2. Create a high-level plan for the solution. 3. Break down the problem into sub tasks. 4. Explain your thought process with justifications. 5. Combine sub task solutions for the main task. 6. Write code to complete the task."
    },
    {
      "role": "user",
      "content": "Write out the conversation so far."
    },
    {
      "role": "assistant",
      "content": ""
    }
  ],
  "options": {
    "temperature": 0,
    "repeat_penalty": 1,
    "repeat_last_n": 64
  },
  "stream": true
}
```

> Here's the conversation so far:
> User: You are a helpful AI assistant.

Which is what is in the modelfile and not the system message I sent.

```
FROM mixtral:32k
TEMPLATE """[INST] {{ if .System }}{{ .System }} {{ end }}{{ .Prompt }} [/INST]{{ .Response }}"""
SYSTEM """You are a helpful AI assistant."""
```

I also tried removing the 1{{ if .System }}1 and it still doesn't work:

```
TEMPLATE """[INST] {{ .System }} {{ .Prompt }} [/INST]{{ .Response }}"""
SYSTEM """You are a helpful AI assistant."""
```

> Here's the conversation so far:
> User: You are a helpful AI assistant.

How can this have gone unnoticed? I only realized something was wrong when I couldn't get Miqu to wrap his code blocks whatever I tried to ask...

**Comment by jukofyork** [comment]
The fix:

`server/images.go`

```
                case "system": 
                        //if currentVars.System != "" {
                        //      if err := writePrompt(); err != nil {
                        //              return "", nil, err
                        //      }
                        //}
                        currentVars.System = msg.Content
```

I'm sorry but I'm not writing any more PRs as I don't have any way to edit Go files other than on Github one commit at a time and I can't make a second folk until the other PR(s) are accepted or rejected... Somebody else will have to fix this.

**Comment by jukofyork** [comment]
Miqu was actually better with the broken system prompt:

```
Write Comments

From lines 323 to 325 of src/eclipse/plugin/aiassistant/utility/Eclipse.java:
public static void saveAllEditors(boolean confirm) {
	Eclipse.getWorkbench().saveAllEditors(confirm);
}
When commenting this source code snippet follow Java best practices for writing comments (such as Javadoc for Java, Doxygen for C++, etc). Clearly and concisely explain the code's purpose and any non-obvious logic/choices - don't just describing code itself! Identify and remove/update any unnecessary, inaccurate, or outdated comments.
```

```
1. Understand the problem:
The problem is to write a comment for the source code snippet that saves all open editors in the Eclipse IDE. The comment should clearly and concisely explain the code's purpose and any non-obvious logic/choices.
2. Create a high-level plan for the solution:
The solution is to write a Javadoc comment that explains the purpose of the method, its parameters, and its return type.
3. Break down the problem into sub tasks:
* Identify the purpose of the method
* Identify the parameters of the method
* Identify the return type of the method
* Write a Javadoc comment that includes this information
4. Explain your thought process with justifications:
The purpose of the method is to save all open editors in the Eclipse IDE. This is a useful feature for users who have multiple files open and want to ensure that all changes are saved before closing the IDE or switching to a different task.
The method takes a single boolean parameter, confirm, which determines whether the user should be prompted to save any unsaved changes before the editors are closed. If confirm is true, the user will be prompted to save any unsaved changes. If confirm is false, any unsaved changes will be discarded without prompting the user.
The method does not have a return type, as it is a void method.
5. Combine sub task solutions for the main task:
Here is the Javadoc comment for the saveAllEditors method:
   
/**
 * Saves all open editors in the Eclipse IDE.
 *
 * @param confirm whether to prompt the user to save any unsaved changes
 *                before closing the editors. If true, the user will be prompted
 *                to save any unsaved changes. If false, any unsaved changes
 *                will be discarded without prompting the user.
 */
public static void saveAllEditors(boolean confirm) {
	Eclipse.getWorkbench().saveAllEditors(confirm);
}

6. Write code to complete the task:
The Javadoc comment above is the completed code for this task.

Note: The original source code snippet did not contain any comments, so there are no unnecessary, inaccurate, or outdated comments to remove or update.
```

:rofl::rofl::rofl::rofl:

**Comment by jukofyork** [comment]
I **think** this works and correctly overrides the modelfile system message iff the first message sent it a "system" role:

```
case "system":
    if currentVars.First == true { // Override iff the very first message.
        currentVars.System = msg.Content // Override the MODELFILE's system message.
        if err := writePrompt(); err != nil {
            return "", nil, err
        }
  }
  // Now  currentVars = PromptVars{} so this case can't be triggered again
  // and "user", "assistant" and post-loop if-statement can't output another system message...
```

but the logic of the whole function really needs looking at closely as it seems really convoluted and error prone...

**Comment by jukofyork** [comment]
> I **think** this works and correctly overrides the modelfile system message iff the first message sent is a "system" role:
> 
> ```
> case "system":
>     if currentVars.First == true { // Override iff the very first message.
>         currentVars.System = msg.Content // Override the MODELFILE's system message.
>         if err := writePrompt(); err != nil {
>             return "", nil, err
>         }
>   }
>   // Now  currentVars = PromptVars{} so this case can't be triggered again
>   // and "user", "assistant" and post-loop if-statement can't output another system message...
> ```
> 
> but the logic of the whole function really needs looking at closely as it seems really convoluted and error prone...

Just want to add this this doesn't work quite as intended either as when you try to use the `.First` variable in the modelfile template it's set false by the time you get to the actual first message if there was a system prompt.

**Comment by jukofyork** [comment]
I'm pretty sure this is the fix:

```
if currentVars.First == true {
    currentVars = PromptVars{
        First:  true,         // Reset to use on next "user" or "assistant" case.
        System: msg.Content,  // Override default with the new system message.
    }
}
```

From `server/images.go`  for context:

```
func (m *Model) ChatPrompt(msgs []api.Message) (string, []api.ImageData, error) {
        // build the prompt from the list of messages
        var prompt strings.Builder
        var currentImages []api.ImageData
        currentVars := PromptVars{
                First:  true,
                System: m.System,
        }

        writePrompt := func() error {
                p, err := Prompt(m.Template, currentVars)
                if err != nil {
                        return err
                }
                prompt.WriteString(p)
                currentVars = PromptVars{}
                return nil
        }

        for _, msg := range msgs {
                switch strings.ToLower(msg.Role) {
                case "system":
                        if currentVars.First == true {
                                currentVars = PromptVars{
                                        First:  true,        // Reset to use on next "user" or "assistant" case.
                                        System: msg.Content, // Override default with the new system message.
                                }
                        }
                case "user":
                        if currentVars.Prompt != "" {
                                if err := writePrompt(); err != nil {
                                        return "", nil, err
                                }
                        }
                        currentVars.Prompt = msg.Content
                        currentImages = msg.Images
                case "assistant":
                        currentVars.Response = msg.Content
                        if err := writePrompt(); err != nil {
                                return "", nil, err
                        }
                default:
                        return "", nil, fmt.Errorf("invalid role: %s, role must be one of [system, user, assistant]", msg.Role)
                }
        }

        // Append the last set of vars if they are non-empty
        if currentVars.Prompt != "" || currentVars.System != "" {
                p, err := m.PreResponsePrompt(currentVars)
                if err != nil {
                        return "", nil, fmt.Errorf("pre-response template: %w", err)
                }
                prompt.WriteString(p)
        }

        return prompt.String(), currentImages, nil
}
```

Basically, if we are given a "system" role message by the API as the first message, reset `currentVars` to the be the same as before the start of the loop:

```
currentVars := PromptVars{
    First:  true,
    System: m.System,
}
```

but replace the original `System: m.System` and with `msg.Content`, and wait for the next iteration of the loop for a "user" or "assistant" case to handle it in the same way as would have happened had no "system" role message been sent via the API.

If the original function is working as expected then this should also work. It should possibly also trigger an error/warning if the "system" role message was not sent first, as the above fix will just silently ignore this...

**Comment by BruceMacD** [comment]
Hi @jukofyork thanks for all the details on this issue. I believe this is fixed in the most recent release of Ollama (v0.1.25) we did some clean-up around this logic recently. So updating should fix this issue.

Here are my testing steps if you'd like to confirm:
1. Run `OLLAMA_DEBUG=1 ollama serve` to start the server with debug logging. This will print the formatted prompt that is being sent to the LLM.
2. Send the request:
```
curl -X POST http://localhost:11434/api/chat -H "Content-Type: application/json" -d '{
  "model": "mistral",
  "messages": [
    {
      "role": "system",
      "content": "You are an AI assistant for the Eclipse IDE. Your objective is to assist users in writing and analyzing source code. Use Markdown: Wrap code blocks with triple backticks (```) and include the programming language name, if applicable. Use single backticks (`) to denote a word or phrase as code. Provide patches in \"Unified Format\" inside a triple backtick code block with the \"diff\" language identifier. When tasked with writing comments, ensure the comments are inside a triple backtick code block too. When tasked with writing code: 1. Understand the problem: constraints, specifications, objective, and edge cases. 2. Create a high-level plan for the solution. 3. Break down the problem into sub tasks. 4. Explain your thought process with justifications. 5. Combine sub task solutions for the main task. 6. Write code to complete the task."
    },
    {
      "role": "user",
      "content": "Write out the conversation so far."
    },
    {
      "role": "assistant",
      "content": ""
    }
  ],
  "options": {
    "temperature": 0,
    "repeat_penalty": 1,
    "repeat_last_n": 64
  },
  "stream": true
}'
```
3. Observe the properly formatted template in the logs:
```
time=2024-02-16T10:12:05.022-04:00 level=DEBUG source=routes.go:1165 msg="chat handler" prompt="[INST] You are an AI assistant for the Eclipse IDE. Your objective is to assist users in writing and analyzing source code. Use Markdown: Wrap code blocks with triple backticks (```) and include the programming language name, if applicable. Use single backticks (`) to denote a word or phrase as code. Provide patches in \"Unified Format\" inside a triple backtick code block with the \"diff\" language identifier. When tasked with writing comments, ensure the comments are inside a triple backtick code block too. When tasked with writing code: 1. Understand the problem: constraints, specifications, objective, and edge cases. 2. Create a high-level plan for the solution. 3. Break down the problem into sub tasks. 4. Explain your thought process with justifications. 5. Combine sub task solutions for the main task. 6. Write code to complete the task. Write out the conversation so far. [/INST]"
```
4. Check the reply is expected.

Resolving this for now, please let me know if the issue persists.

**Comment by BruceMacD** [comment]
I found a couple more issues that could have been the source while testing the other issue you commented in (#2942), working on fixing them at the moment for the next release. I'll let you know what the specifics are there too.
